cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/util/MultiThreadedTestHelper.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2016 IBM Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+package com.cloudant.sync.util;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A simple test abstraction that executes a callable from the specified number of threads as close
+ * to simultaneously as possible. The callable is wrapped in a runnable and blocked behind a latch.
+ * The latch counts down as each thread is ready and then there is a final countdown to unlatch all
+ * the threads at once.
+ * <p>
+ * To use this class extend it specifying the number of threads and implementing getCallable.
+ * </p>
+ * <p>
+ * The default assertions include checking that all the threads executed and there were no
+ * exceptions and that the number of results were correct. Additional assertions are added by
+ * implementing the doAssertions method.
+ * </p>
+ * <p>
+ * Call run to start the test.
+ * </p>
+ *
+ * @param <T> the type of result from the Callable
+ */
+public abstract class MultiThreadedTestHelper<T> {
+
+    private final int numberOfThreads;
+    private final List<Thread> threads;
+    protected final List<T> results;
+    protected final List<Throwable> exceptions;
+    final AtomicBoolean failedToExecuteAllThreads = new AtomicBoolean(false);
+    final CountDownLatch latch;
+
+
+    public MultiThreadedTestHelper(int numberOfThreads) {
+        this.numberOfThreads = numberOfThreads;
+        this.latch = new CountDownLatch(numberOfThreads + 1);
+        this.threads = new ArrayList<Thread>(numberOfThreads);
+        this.results = Collections.synchronizedList(new ArrayList<T>(numberOfThreads));
+        this.exceptions = Collections.synchronizedList(new ArrayList<Throwable>(numberOfThreads));
+        for (int i = 0; i < numberOfThreads; i++) {
+            final Callable<T> callable = getCallable();
+            threads.add(new Thread(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        latch.countDown();
+                        if (latch.await(20, TimeUnit.SECONDS)) {
```

Comment:
This might benefit from a comment like "wait for all theads to start" just to be clear that we're waiting on that and not the actual callable - I know you explained this to me before.